oshmem/shmem/c/shmem_mcs_lock.c

```diff
@@ -0,0 +1,256 @@
+/**
+ * @file
+ */
+#include "oshmem_config.h"
+
+#include "oshmem/constants.h"
+#include "oshmem/include/shmem.h"
+#include "oshmem/runtime/params.h"
+#include "oshmem/runtime/runtime.h"
+#include <stdlib.h>
+#include <memory.h>
+
+#include "oshmem/shmem/shmem_api_logger.h"
+#include "oshmem/shmem/shmem_lock.h"
+#include "oshmem/mca/memheap/memheap.h"
+#include "oshmem/mca/memheap/base/base.h"
+#include "oshmem/mca/atomic/atomic.h"
+
+#define OPAL_BITWISE_SIZEOF_LONG (SIZEOF_LONG * 8)
+
+
+#define MCSQ_TAIL_OWNER(lock_ptr)\
+	(((uintptr_t)(lock_ptr) / sizeof(long)) % shmem_n_pes())
+
+/*
+ * Use basic MCS distributed lock algorithm for lock
+ */
+struct lock_t {
+	/* has meaning only on MCSQ_TAIL OWNER */
+	int tail;
+	/* has meaning on all PEs */
+	/**
+	 * The next pointer is a combination of
+	 * the PE ID and the wait signal
+	 */
+	int next;
+};
+typedef struct lock_t lock_t;
+#define NEXT_MASK	0x7FFFFFFFU
+#define SIGNAL_MASK	0x80000000U //Wait signal mask
+#define NEXT(A)		(A & NEXT_MASK)
+#define GET_PE(P)	(P & NEXT_MASK) // Improve readability
+#define SIGNAL(A)	(A & SIGNAL_MASK)
+#define SET_SIGNAL(A)	(A | SIGNAL_MASK)
+
+void
+_shmem_mcs_set_lock(long *lockp)
+{
+
+	lock_t	*lock =	(lock_t *) lockp;
+	int	mcs_tail_owner = MCSQ_TAIL_OWNER(lock);
+	int	new_tail_req = 0;
+	int	*tail = &(lock->tail);
+	int	*next = &(lock->next);
+	int	my_pe = shmem_my_pe();
+	int	curr = 0;
+	int	out_value = 0;
+	int	prev_tail = 0;
+	int	prev_tailpe = 0;
+	int	tval = 0;
+	int	tmp_val = 0;
+	int	retv = 0;
+	uint64_t value_tmp = 0;
+
+	RUNTIME_CHECK_INIT();
+	/**
+	 *  Initializing next pointer to next mask
+	 *  Done atomically to avoid races as NEXT pointer
+	 *  can be modified by other PEs while acquiring or
+	 *  releasing it.
+	 */
+	/**
+	 * Can make this to be shmem_atomic_set to be safe in non-cc architecutres
+	 * has an impact on performance
+	 */
+	value_tmp = NEXT_MASK;
+	out_value = NEXT_MASK;
+	retv = MCA_ATOMIC_CALL(swap(oshmem_ctx_default, (void*)next,
+				(void*)&out_value, value_tmp, sizeof(int),
+				my_pe));
+	RUNTIME_CHECK_RC(retv);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+	/** Signal for setting lock */
+	new_tail_req = SET_SIGNAL(my_pe);
+	/**
+	 * Swap and make me the new tail and update in tail owner
+	 * Get the previous tail PE.
+	 */
+	retv = MCA_ATOMIC_CALL(swap(oshmem_ctx_default,
+			(void *)tail,
+			(void*)&prev_tail,
+			OSHMEM_ATOMIC_PTR_2_INT(&new_tail_req, sizeof(new_tail_req)),
+			sizeof(int), mcs_tail_owner));
+	RUNTIME_CHECK_RC(retv);
+	prev_tailpe	= GET_PE(prev_tail);
+
+	if (SIGNAL(prev_tail)) {
+		/**
+		  * Someone else has got the lock before this PE
+		  * Adding this PE to the previous tail PE's Next pointer
+		  * Substract the SIGNAL Bit to avoid changing it.
+		  **/
+		tmp_val = my_pe - NEXT_MASK;
+		retv = MCA_ATOMIC_CALL(add(oshmem_ctx_default, (void*)next, tmp_val,
+							 sizeof(int), prev_tailpe));
+		RUNTIME_CHECK_RC(retv);
+
+		/**
+		 *	This value to be changed eventually by predecessor
```

Comment:
alignment

Reply:
I have redone the alignment for the whole file now, hopefully this patch fixes the alignment issues.